src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+#include <string>
+#include <sstream>
+#include <time.h>
+#include <iomanip> 
+#include <iostream>
+#include <cstdio>
+#include <filesystem>
 
-#include "local_cache.h"
+#define NT_SUCCESS(Status)          (((NTSTATUS)(Status)) >= 0)
 
-//
-// The Windows implementation of the local cache is not yet complete.
-// All methods noop, so we'll never cache any data locally.
-//
+#define STATUS_UNSUCCESSFUL         ((NTSTATUS)0xC0000001L)
 
-void local_cache_clear()
+#define MAX_RETRY					10000
+#define SLEEP_RETRY_MS				15
+
+constexpr uint16_t CACHE_V1 = 1;
+
+constexpr _locale_t NULL_LOCALE = reinterpret_cast<_locale_t>(0);
+
+static std::string g_cache_dirname;
+
+static void throw_if(bool should_throw, const std::string& error)
+{
+    if (should_throw)
+    {
+        throw std::runtime_error(error);
+    }
+}
+
+static void throw_errno(const std::string& description, int err)
 {
-    // noop
+
+    _locale_t locale = _create_locale(LC_ALL, "POSIX");
+    if (locale == NULL_LOCALE)
+    {
+        throw std::runtime_error("Unable to allocate locale: " + std::to_string(errno));
+    }
+    std::string errno_string = strerror(err);
+    _free_locale(locale);
+
+    throw std::runtime_error(description + ": " + errno_string);
 }
 
-void local_cache_add(
-    const std::string& id,
-    time_t expiry,
-    size_t data_size,
-    const void* data)
+static void throw_errno(const std::string& description)
 {
-    (void)id;
-    (void)expiry;
-    (void)data_size;
-    (void)data;
+    throw_errno(description, errno);
+}
+
+struct CacheEntryHeaderV1 {
+    uint16_t version;   // The version of the cache header
+    time_t expiry;      // expiration time of this cache item
+};
+
+static void make_dir(const std::string& dirname)
+{
+    struct _stat buf {};
+    int rc = _stat(dirname.c_str(), &buf);
+
+    if (rc == 0)
+    {
+        if ((buf.st_mode & _S_IFDIR) > 0)
+        {
+            return;
+        }
+
+        throw std::runtime_error(dirname + " already exists, and is not a directory.");
+    }
+
+    wchar_t dirName[64];
+    std::mbstowcs(dirName, dirname.c_str(), strlen(dirname.c_str()) + 1);//Plus null
+
+    rc = CreateDirectory(dirName, NULL);
+    if (rc != 0)
+    {
+        throw_errno("Error creating directory '" + dirname + "'");
+    }
+}
+
+static void init_callback()
+{
+    const char * env_home = ::getenv("APPDATA");
+    const char * env_azdcap_cache = ::getenv("AZDCAP_CACHE");
+    const std::string application_name("\\.az-dcap-client");
+
+    std::string dirname;
+
+    if (env_azdcap_cache != 0 && (strcmp(env_azdcap_cache, "") != 0))
+    {
+        dirname = env_azdcap_cache;
+    }
+    else if (env_home != 0 && (strcmp(env_home, "") != 0))
+    {
+        dirname = std::string(env_home);
+    }
+    else
+    {
+        // Throwing exception if the expected HOME
+        // environment variable is not defined.
+        throw std::runtime_error("HOME and AZDCAPCACHE environment variables not defined");
```

Comment:
This says it depends on an environment variable "HOME", but instead it depends on APPDATA.

Reply:
Changed